File: python/file_loader.py
import csv
import os
from scheduler import EmployeeScheduler


def load_from_file(scheduler: EmployeeScheduler, filename: str) -> bool:
    print(f"\nLoading data from {filename}...")
    
    if not os.path.exists(filename):
        print(f"Error: File '{filename}' not found.")
        return False
    
    try:
        with open(filename, 'r') as file:
            csv_reader = csv.reader(file)
            
            employees_added = set()
            preferences_count = 0
            
            for line_num, row in enumerate(csv_reader, 1):
                if not row or not any(row):
                    continue
                
                if row[0].strip().startswith('#'):
                    continue
                
                if len(row) < 3:
                    print(f"Warning: Line {line_num} has insufficient data, skipping")
                    continue
                
                employee = row[0].strip()
                day = row[1].strip()
                shifts = [s.strip() for s in row[2:] if s.strip()]
                
                if day not in scheduler.DAYS:
                    print(f"Warning: Line {line_num} has invalid day '{day}', skipping")
                    continue
                
                valid_shifts = [s for s in shifts if s in scheduler.SHIFTS]
                if not valid_shifts:
                    print(f"Warning: Line {line_num} has no valid shifts, skipping")
                    continue
                
                if employee not in employees_added:
                    scheduler.add_employee(employee)
                    employees_added.add(employee)
                
                scheduler.preferences[employee][day] = valid_shifts
                preferences_count += 1
                print(f"  Loaded: {employee} - {day}: {', '.join(valid_shifts)}")
            
            print(f"\n✓ Successfully loaded {len(employees_added)} employees with {preferences_count} preferences")
            return True
            
    except Exception as e:
        print(f"Error reading file: {e}")
        return False

File: java/src/util/FileLoader.java
package util;

import java.io.*;
import java.util.*;
import scheduler.Scheduler;

public class FileLoader {
    
    public static boolean loadFromFile(Scheduler scheduler, String filename) {
        System.out.println("\nLoading data from " + filename + "...");
        
        File file = new File(filename);
        if (!file.exists()) {
            System.out.println("Error: File '" + filename + "' not found.");
            return false;
        }
        
        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            Set<String> employeesAdded = new HashSet<>();
            int preferencesCount = 0;
            int lineNum = 0;
            String line;
            
            while ((line = reader.readLine()) != null) {
                lineNum++;
                
                if (line.trim().isEmpty()) continue;
                if (line.trim().startsWith("#")) continue;
                
                String[] parts = line.split(",");
                
                if (parts.length < 3) {
                    System.out.println("Warning: Line " + lineNum + " has insufficient data, skipping");
                    continue;
                }
                
                String employee = parts[0].trim();
                String day = parts[1].trim();
                
                List<String> shifts = new ArrayList<>();
                for (int i = 2; i < parts.length; i++) {
                    String shift = parts[i].trim();
                    if (!shift.isEmpty()) {
                        shifts.add(shift);
                    }
                }
                
                if (!Arrays.asList(Scheduler.DAYS).contains(day)) {
                    System.out.println("Warning: Line " + lineNum + " has invalid day '" + day + "', skipping");
                    continue;
                }
                
                List<String> validShifts = new ArrayList<>();
                for (String shift : shifts) {
                    if (Arrays.asList(Scheduler.SHIFTS).contains(shift)) {
                        validShifts.add(shift);
                    }
                }
                
                if (validShifts.isEmpty()) {
                    System.out.println("Warning: Line " + lineNum + " has no valid shifts, skipping");
                    continue;
                }
                
                if (!employeesAdded.contains(employee)) {
                    scheduler.addEmployee(employee);
                    employeesAdded.add(employee);
                }
                
                scheduler.preferences.get(employee).put(day, validShifts);
                preferencesCount++;
                System.out.println("  Loaded: " + employee + " - " + day + ": " + 
                                 String.join(", ", validShifts));
            }
            
            System.out.println("\n✓ Successfully loaded " + employeesAdded.size() + 
                             " employees with " + preferencesCount + " preferences");
            return true;
            
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return false;
        }
    }
}
